trin-core/src/portalnet/overlay/protocol.rs

```diff
@@ -243,61 +223,87 @@ impl OverlayProtocol {
             .collect()
     }
 
-    pub async fn send_ping(
-        &self,
-        data_radius: U256,
-        enr: Enr,
-        protocol: ProtocolKind,
-    ) -> Result<Vec<u8>, String> {
+    /// Sends a `Ping` request to `enr`.
+    pub async fn send_ping(&self, enr: Enr) -> Result<Pong, String> {
         let enr_seq = self.discovery.read().await.local_enr().seq();
-        let msg = Ping {
+        let request = Ping {
             enr_seq,
-            data_radius,
+            data_radius: self.data_radius().await,
         };
-        self.discovery
-            .read()
-            .await
-            .send_talkreq(
-                enr,
-                protocol.to_string(),
-                Message::Request(Request::Ping(msg)).to_bytes(),
-            )
-            .await
+
+        // Send a request through the overlay service.
+        let (tx, rx) = oneshot::channel();
+        let overlay_request = OverlayRequest {
+            request: Request::Ping(request),
+            direction: RequestDirection::Outgoing { destination: enr },
+            responder: Some(tx),
+        };
+        if let Err(_) = self.request_tx.send(overlay_request) {
+            return Err("Receiver half of channel dropped, unable to submit request".to_owned());
+        }
+
+        // Wait on the response.
+        match rx.await {
```

Comment:
Not sure how I feel about matching results wrapped around results all simultaneously. I'd recommend handling each level of result one after the other.